lib/Models/GeoJsonCatalogItem.ts

```diff
@@ -0,0 +1,674 @@
+                    });
+                }
+                jsonPromise = loadZipFile(proxyCatalogItemUrl(this, url, "1d"));
+            } else {
+                jsonPromise = loadJson(proxyCatalogItemUrl(this, url, "1d"));
+            }
+
+            return jsonPromise
+                .then((data: any) => {
+                    return this.setTraitsFromData(data);
+                })
+                .otherwise((e?: Error) => {
+                    if (e instanceof TerriaError) {
+                        throw e;
+                    } else {
+                        throw new TerriaError({
+                            sender: this,
+                            title: "Could not load JSON",
+                            message: `An error occurred while retrieving JSON data from the provided link.<p>If you entered the link manually, please verify that the link is correct.</p><p>This error may also indicate that the server does not support <a href="http://enable-cors.org/" target="_blank">CORS</a>. If this is your server, verify that CORS is enabled and enable it if it is not.  If you do not control the server, please contact the administrator of the server and ask them to enable CORS. Or, contact the ${appName} team by emailing <a href="mailto:${supportEmail}">${supportEmail}</a> and ask us to add this server to the list of non-CORS-supporting servers that may be proxied by ${appName} itself.</p><p>If you did not enter this link manually, this error may indicate that the data source you\'re trying to add is temporarily unavailable or there is a problem with your internet connection.  Try adding the data source again, and if the problem persists, please report it by sending an email to <a href="mailto:${supportEmail}">${supportEmail}</a>.</p>`
+                        });
+                    }
+                });
+        }
+    }
+
+    setTraitsFromData(geoJson: any): Promise<void> {
+        // If this GeoJSON data is an object literal with a single property, treat that
+        // property as the name of the data source, and the property's value as the
+        // actual GeoJSON.
+        var numProperties = 0;
+        var propertyName;
+        for (propertyName in geoJson) {
+            if (geoJson.hasOwnProperty(propertyName)) {
+                ++numProperties;
+                if (numProperties > 1) {
+                    break; // no need to count past 2 properties.
+                }
+            }
+        }
+
+        var name;
+        if (numProperties === 1 && isDefined(propertyName)) {
+            name = propertyName;
+            geoJson = geoJson[propertyName];
+            this.name = name;
```

Comment:
Wherever possible, we should think in terms of "this property is a function of this one/these" instead of "when that happens, this happens." So I'd try something like this:

* `LoadGeoJsonStratum` has a private, observable property called `_loadedGeoJson`, which is asynchronously set by `loadGeoJson`. This is the cause/effect bit, but we make it as small as possible.
* `name` is a computed property that reads `_loadedGeoJson` and uses the logic in this function to either return undefined (if there is more than one property), or the GeoJSON's outer property name (if there is exactly one outer property).
* `geoJsonData` is a computed property that returns `_loadedGeoJson` `geoJson[propertyName]` if there is exactly one outer property in `_loadedGeoJson`, and `_loadedGeoJson` itself otherwise.

Reply:
The common logic in the last two bullets can even be moved to a separate private computed. Values of computed observables are cached and only recomputed when something they depend on changes (not that that should make much difference here).

Reply:
Understood the gist of the comment. But the suggested change to `@compute geoJsonData` from the unwrapped `_loadedGeoJson` might not be doable since we also need to `reprojectToGeographic()` which is an asynchronous call that will turn `geoJsonData` into a `Promise`.

Reply:
Ah I see, that's a good point. I _think_ it can be handled again by deferring reprojection until `mapItems` is needed, and then constructing and returning the `GeoJsonDataSource` immediately, then doing the reprojection, and when the reproject promise resolves call `load` with the reprojected data. Does that make sense?